prometheus/histogram.go

```diff
@@ -1575,3 +1593,152 @@ func addAndResetCounts(hot, cold *histogramCounts) {
 	atomic.AddUint64(&hot.nativeHistogramZeroBucket, atomic.LoadUint64(&cold.nativeHistogramZeroBucket))
 	atomic.StoreUint64(&cold.nativeHistogramZeroBucket, 0)
 }
+
+type nativeExemplars struct {
+	exemplars                       map[*dto.Exemplar]*totalItem
+	nativeHistogramExemplarTTL      time.Duration
+	nativeHistogramMaxExemplarCount uint32
+	timeList                        *timeItem
+	logarithmList                   *logarithmItem
+}
+
+type totalItem struct {
+	timeItem      *timeItem
+	logarithmItem *logarithmItem
+}
+
+func newNativeExemplars(ttl time.Duration, count uint32) *nativeExemplars {
+	return &nativeExemplars{
+		nativeHistogramExemplarTTL:      ttl,
+		nativeHistogramMaxExemplarCount: count,
+		exemplars:                       make(map[*dto.Exemplar]*totalItem),
+		timeList:                        &timeItem{},
+		logarithmList:                   &logarithmItem{},
+	}
+}
+
+func (n *nativeExemplars) addExemplar(e *dto.Exemplar) {
```

Comment:
So the problem here is that this isn't concurrency safe. Note how the exemplars for classic exemplars are done (with a `[]atomic.Value`).

In the worst case, you could put everything behind a mutex, but that's quite slow. One might argue that `ObserveWithExemplar` shouldn't be called with high frequency anyway, and the mutex is OK for it (but that needs to be documented, should we go down that path).

I'm not sure if there is a feasible way to do something faster than the full mutex protection. For classic buckets, it's easy to know which exemplar to replace, but with the algorithm here, it's a much more complex decision.

Maybe let's do the mutex approach first and then try something more sophisticated if need arises.

Reply:
added a `sync.Mutex` to `nativeExemplars` struct